File: app/platforms/instagram.py
import httpx
from app.utils.logger import logger

async def get_auth_url(client_id: str, redirect_uri: str, state: str, scopes: list):
    params = {
        "client_id": client_id,
        "redirect_uri": redirect_uri,
        "state": state,
        "response_type": "code",
        "scope": ",".join(scopes)
    }
    return f"https://www.facebook.com/v19.0/dialog/oauth?{'&'.join([f'{k}={v}' for k,v in params.items()])}"

async def exchange_code(client_id: str, client_secret: str, redirect_uri: str, code: str):
    async with httpx.AsyncClient() as client:
        res = await client.get(
            "https://graph.facebook.com/v19.0/oauth/access_token",
            params={
                "client_id": client_id,
                "client_secret": client_secret,
                "redirect_uri": redirect_uri,
                "code": code
            }
        )
        res.raise_for_status()
        return res.json()

import os

async def publish_image(access_token: str, ig_user_id: str, image_url: str, caption: str, local_path: str = None):
    async with httpx.AsyncClient() as client:
        # 1. Create container
        params = {
            "caption": caption,
            "access_token": access_token
        }
        files = None
        if local_path and os.path.exists(local_path):
            files = {"image_url": open(local_path, "rb")}
        else:
            params["image_url"] = image_url

        res = await client.post(
            f"https://graph.facebook.com/v19.0/{ig_user_id}/media",
            params=params if not files else {"access_token": access_token, "caption": caption},
            files=files
        )
        res.raise_for_status()
        container_id = res.json().get("id")

        # 2. Publish
        pub_res = await client.post(
            f"https://graph.facebook.com/v19.0/{ig_user_id}/media_publish",
            params={"creation_id": container_id, "access_token": access_token}
        )
        pub_res.raise_for_status()
        return pub_res.json()

async def publish_video(access_token: str, ig_user_id: str, video_url: str, caption: str, local_path: str = None):
    async with httpx.AsyncClient() as client:
        # 1. Create container
        params = {
            "media_type": "VIDEO",
            "caption": caption,
            "access_token": access_token
        }
        files = None
        if local_path and os.path.exists(local_path):
            # Instagram requires video to be uploaded as 'video_url' normally, 
            # but 'video_url' in multipart form is also supported for some endpoints.
            files = {"video_url": open(local_path, "rb")}
        else:
            params["video_url"] = video_url

        res = await client.post(
            f"https://graph.facebook.com/v19.0/{ig_user_id}/media",
            params=params if not files else {"access_token": access_token, "media_type": "VIDEO", "caption": caption},
            files=files
        )
        res.raise_for_status()
        container_id = res.json().get("id")

        # 2. Poll
        import asyncio
        max_retries = 30
        for _ in range(max_retries):
            status_res = await client.get(
                f"https://graph.facebook.com/v19.0/{container_id}",
                params={"fields": "status_code", "access_token": access_token}
            )
            data = status_res.json()
            if data.get("status_code") == "FINISHED":
                break
            if data.get("status_code") == "ERROR":
                raise Exception(f"Instagram media processing failed: {data.get('status_msg', 'Unknown Error')}")
            await asyncio.sleep(5)

        # 3. Publish
        pub_res = await client.post(
            f"https://graph.facebook.com/v19.0/{ig_user_id}/media_publish",
            params={"creation_id": container_id, "access_token": access_token}
        )
        pub_res.raise_for_status()
        return pub_res.json()

async def publish_carousel(access_token: str, ig_user_id: str, media_urls: list, caption: str):
    """
    Publish a carousel (multi-item) post to Instagram.
    media_urls: List of dicts with {url, is_video}
    """
    async with httpx.AsyncClient() as client:
        # 1. Create individual item containers
        item_ids = []
        for item in media_urls:
            params = {
                "access_token": access_token,
                "is_carousel_item": "true",
                "caption": caption # Caption is technically ignored for carousel items but good to have
            }
            if item.get("is_video"):
                params["media_type"] = "VIDEO"
                params["video_url"] = item["url"]
            else:
                params["image_url"] = item["url"]
                
            res = await client.post(f"https://graph.facebook.com/v19.0/{ig_user_id}/media", params=params)
            res.raise_for_status()
            item_ids.append(res.json().get("id"))

        # 2. Wait for items to process (especially videos)
        import asyncio
        for container_id in item_ids:
            max_retries = 20
            for _ in range(max_retries):
                status_res = await client.get(
                    f"https://graph.facebook.com/v19.0/{container_id}",
                    params={"fields": "status_code", "access_token": access_token}
                )
                if status_res.json().get("status_code") == "FINISHED":
                    break
                await asyncio.sleep(3)

        # 3. Create carousel container
        carousel_res = await client.post(
            f"https://graph.facebook.com/v19.0/{ig_user_id}/media",
            params={
                "media_type": "CAROUSEL",
                "children": ",".join(item_ids),
                "caption": caption,
                "access_token": access_token
            }
        )
        carousel_res.raise_for_status()
        carousel_id = carousel_res.json().get("id")

        # 4. Publish carousel
        pub_res = await client.post(
            f"https://graph.facebook.com/v19.0/{ig_user_id}/media_publish",
            params={"creation_id": carousel_id, "access_token": access_token}
        )
        pub_res.raise_for_status()
        return pub_res.json()

async def get_me(access_token: str):
    async with httpx.AsyncClient() as client:
        res = await client.get(
            "https://graph.facebook.com/v19.0/me",
            params={
                "fields": "id,name,username,account_type,media_count",
                "access_token": access_token
            }
        )
        res.raise_for_status()
        return res.json()

File: src/platforms/instagram.js
const axios = require('axios');
const { AppError } = require('../utils/errors');

/**
 * Instagram Graph API requires:
 * - Instagram Business/Creator account
 * - Connected to a Facebook Page
 *
 * Typical posting flow:
 * 1) /{ig-user-id}/media (create container)
 * 2) /{ig-user-id}/media_publish
 *
 * This adapter includes a minimal skeleton. You will need media URLs (publicly accessible HTTPS),
 * and proper permissions and app review.
 */

function getConfig() {
  const appId = process.env.FACEBOOK_APP_ID;
  const appSecret = process.env.FACEBOOK_APP_SECRET;
  if (!appId || !appSecret) throw new Error('FACEBOOK_APP_ID/FACEBOOK_APP_SECRET missing (Instagram uses Meta app)');
  const scopes = (process.env.INSTAGRAM_SCOPES || '').split(',').map(s => s.trim()).filter(Boolean);
  return { appId, appSecret, scopes };
}

function getAuthUrl({ redirectUri, state, codeChallenge }) {
  const { appId, scopes } = getConfig();
  const params = new URLSearchParams({
    client_id: appId,
    redirect_uri: redirectUri,
    state,
    response_type: 'code',
    scope: scopes.join(',')
  });
  params.set('code_challenge', codeChallenge);
  params.set('code_challenge_method', 'S256');
  return `https://www.facebook.com/v24.0/dialog/oauth?${params.toString()}`;
}

async function exchangeCode({ code, redirectUri }) {
  const { appId, appSecret } = getConfig();
  const params = new URLSearchParams({
    client_id: appId,
    client_secret: appSecret,
    redirect_uri: redirectUri,
    code
  });
  const res = await axios.get(`https://graph.facebook.com/v24.0/oauth/access_token?${params.toString()}`);
  return res.data;
}

async function refreshToken() {
  throw new AppError('Instagram refresh is not implemented (depends on token type)', 501, 'instagram_refresh_not_implemented');
}

async function getProfile({ accessToken }) {
  // IG Graph does not expose IG user ID directly from /me; you need to traverse:
  // /me/accounts -> /{page-id}?fields=instagram_business_account
  const pagesRes = await axios.get('https://graph.facebook.com/v24.0/me/accounts', {
    params: { access_token: accessToken }
  });
  const pages = pagesRes.data?.data || [];
  const firstPage = pages[0];
  if (!firstPage?.id) throw new AppError('No Facebook Page found (required for IG Graph)', 400, 'instagram_no_page');

  const pageRes = await axios.get(`https://graph.facebook.com/v24.0/${firstPage.id}`, {
    params: { access_token: accessToken, fields: 'instagram_business_account,name' }
  });
  const ig = pageRes.data?.instagram_business_account;
  if (!ig?.id) throw new AppError('No Instagram Business/Creator account linked to this Page', 400, 'instagram_not_linked');

  return {
    accountId: ig.id,
    displayName: pageRes.data?.name,
    rawProfile: { pages: pagesRes.data, page: pageRes.data }
  };
}

async function post({ accessToken, content, media }) {
  // For now, require a single image URL in media[0].
  const imageUrl = media?.[0];
  if (!imageUrl) throw new AppError('Instagram posting requires media[0] as a public image URL', 400, 'instagram_media_required');

  // NOTE: We need the ig-user-id (stored as accountId)
  throw new AppError('Instagram publishing endpoint requires accountId; use /post/instagram with accountId from linked accounts', 400, 'instagram_account_required');
}

async function publishImage({ accessToken, igUserId, caption, imageUrl }) {
  const createRes = await axios.post(`https://graph.facebook.com/v24.0/${igUserId}/media`, null, {
    params: { image_url: imageUrl, caption, access_token: accessToken }
  });
  const containerId = createRes.data?.id;
  if (!containerId) throw new AppError('Failed to create IG media container', 502, 'instagram_container_failed', createRes.data);

  const pubRes = await axios.post(`https://graph.facebook.com/v24.0/${igUserId}/media_publish`, null, {
    params: { creation_id: containerId, access_token: accessToken }
  });
  return { id: pubRes.data?.id, raw: { createRes: createRes.data, publishRes: pubRes.data } };
}

function mapAxiosError(err) {
  const status = err.response?.status;
  if (status === 401) return new AppError('Instagram auth failed', 401, 'instagram_auth_failed');
  if (status === 429) return new AppError('Instagram rate limit', 429, 'instagram_rate_limited', err.response?.data);
  return new AppError('Instagram API error', status || 502, 'instagram_api_error', err.response?.data || err.message);
}

module.exports = {
  platform: 'instagram',
  getAuthUrl,
  exchangeCode,
  refreshToken,
  getProfile,
  post,
  publishImage,
  mapAxiosError
};
